File: Evolutionary Algorithms/Python Codes/DE.py
# -*- coding: utf-8 -*-
"""
Created on Thu Sep  7 03:51:51 2023

@author: msi-1
"""

import numpy as np

def sphere_func(x):
    return np.sum(x**2)

def differential_evolution(pop_size, num_vars, lb, ub, num_generations, F, CR):
    # Initialize population
    population = np.random.uniform(lb, ub, (pop_size, num_vars))
    
    # Main loop
    for gen in range(num_generations):
        # Evaluate fitness
        fitness = np.apply_along_axis(sphere_func, 1, population)
        best_fitness = np.min(fitness)
        
        # Create offspring population
        offspring = np.zeros_like(population)
        for i in range(pop_size):
            # Select three distinct individuals from the population
            idx = np.random.choice(pop_size, 3, replace=False)
            x1, x2, x3 = population[idx]
            
            # Mutation
            v = x1 + F * (x2 - x3)
            
            # Crossover
            mask = np.random.rand(num_vars) < CR
            u = np.where(mask, v, population[i])
            
            # Selection
            if sphere_func(u) <= fitness[i]:
                offspring[i] = u
            else:
                offspring[i] = population[i]
        
        # Update population
        population = offspring
        
        # Display current best fitness
        print(f'Generation {gen}: Best Fitness = {best_fitness}')
    
    # Final evaluation
    final_fitness = np.apply_along_axis(sphere_func, 1, population)
    best_solution = population[np.argmin(final_fitness)]
    best_fitness = np.min(final_fitness)
    print(f'Final Best Fitness = {best_fitness}')
    print('Best Solution:')
    print(best_solution)

# DE Parameters
pop_size = 50
num_vars = 3
lb = -100
ub = 100
num_generations = 1000
F = 0.8  # Scaling factor
CR = 0.9  # Crossover rate

# Run DE algorithm
differential_evolution(pop_size, num_vars, lb, ub, num_generations, F, CR)

File: Evolutionary Algorithms/MATLAB Codes/DE.m
clc;
clear all;

% Differential Evolution Parameters
pop_size = 50;
num_vars = 30;
lb = -100;
ub = 100;
num_generations = 1000;
F = 0.5;  % Differential weight
CR = 0.7;  % Crossover probability

% Initialize population
population = lb + (ub - lb) * rand(pop_size, num_vars);
trial_population = zeros(pop_size, num_vars);

% Evaluate initial population
fitness = zeros(pop_size, 1);
for i = 1:pop_size
    fitness(i) = sphere_func(population(i, :));
end

% Main loop
for gen = 1:num_generations
    
    for i = 1:pop_size
        % Mutation
        indices = setdiff(1:pop_size, i);  % Exclude the current index
        selected_indices = randsample(indices, 3);  % Randomly select three distinct indices
        
        a = population(selected_indices(1), :);
        b = population(selected_indices(2), :);
        c = population(selected_indices(3), :);
        
        mutant_vector = a + F * (b - c);
        
        % Crossover
        j_rand = randi([1, num_vars]);
        for j = 1:num_vars
            if rand() <= CR || j == j_rand
                trial_population(i, j) = mutant_vector(j);
            else
                trial_population(i, j) = population(i, j);
            end
        end
    end
    
    % Selection
    for i = 1:pop_size
        trial_fitness = sphere_func(trial_population(i, :));
        if trial_fitness < fitness(i)
            population(i, :) = trial_population(i, :);
            fitness(i) = trial_fitness;
        end
    end
    
    % Display current best fitness
    best_fitness = min(fitness);
    fprintf('Generation %d: Best Fitness = %f\n', gen, best_fitness);
end

% Final evaluation
best_solution = population(fitness == min(fitness), :);
fprintf('Final Best Fitness = %f\n', best_fitness);
disp('Best Solution:');
disp(best_solution);
